s3transfer/utils.py

```diff
@@ -241,3 +253,47 @@ def __iter__(self):
         # already exhausted the stream so iterating over the file immediately
         # stops, which is what we're simulating here.
         return iter([])
+
+
+class StreamReaderProgress(object):
+    """Wrapper for a read only stream that adds progress callbacks."""
+    def __init__(self, stream, callbacks=None):
+        self._stream = stream
+        self._callbacks = callbacks
+        if callbacks is None:
+            self._callbacks = []
+
+    def read(self, *args, **kwargs):
+        value = self._stream.read(*args, **kwargs)
+        for callback in self._callbacks:
+            callback(bytes_transferred=len(value))
+        return value
+
+
+class ShutdownQueue(queue.Queue):
+    """A queue implementation that can be shutdown.
+    Shutting down a queue means that this class adds a
+    trigger_shutdown method that will trigger all subsequent
+    calls to put() to fail with a ``QueueShutdownError``.
+    It purposefully deviates from queue.Queue, and is *not* meant
+    to be a drop in replacement for ``queue.Queue``.
+    """
+    def _init(self, maxsize):
+        self._shutdown = False
+        self._shutdown_lock = threading.Lock()
+        # queue.Queue is an old style class so we don't use super().
+        return queue.Queue._init(self, maxsize)
+
+    def trigger_shutdown(self):
+        with self._shutdown_lock:
+            self._shutdown = True
+            logger.debug("The IO queue is now shutdown.")
+
+    def put(self, item):
+        # Note: this is not sufficient, it's still possible to deadlock!
```

Comment:
I am actually not sure. I just copied this utility from boto3: https://github.com/boto/boto3/blob/8e89d3d023b40ca3fb6f0424efd631f36a8f2533/boto3/s3/transfer.py#L442 comment and all. I am actually not sure how relevant the comment is now. I do not see how it can deadlock. Do you remember? It looks like you were the one to add the note in the first place.